### ISSUE TITLE
Use property decorators for all lifecycle hooks instead of interfaces

### ISSUE THREAD

**Issue by IgorMinar** [angular/angular#5036]
Currently Component and routable component lifecycles are defined via interfaces (e.g. `OnInit`).

This is problematic because the interfaces have no runtime representation in JavaScript. So at runtime, we just check if the method defined in the interface (e.g. `onInit`) exists on the component instance and if it does we call it. If a component has a method `onInit` that is not meant to be a lifecycle hook, we'll still call it.

These name collisions can result in surprising results, and if we add new lifecycle hooks, we can end up breaking existing components.

For these reasons, I propose that we move away from interfaces and use property decorators instead.

Before:
```
import {Component, OnInit} from 'angular2/angular2';

@Component({
  selector: 'test-foo'
})
class TestFoo implements OnInit {
  onInit() { ... }
}
```

After:
```
import {Component, OnInit} from 'angular2/angular2';

@Component({
  selector: 'test-foo'
})
class TestFoo {
  @OnInit() onInit() { ... }
}
```

Dart notes: this api can be implemented via transformers for production and via mirrors for development mode.

Migration notes: I'm not sure how to deal with migration just yet because we are going to reuse the same symbol for different purpose. there might be no good way to provide a transitional path, which would argue for making this change asap.

//cc: @mhevery @vsavkin @tbosch @btford

**Comment by btford** [comment]
+1

**Comment by gdi2290** [comment]
:+1:

**Comment by mhevery** [comment]
@IgorMinar I am fine with this proposal with decorators. I would like to see how the code will look in ES5 where there are no decorators. I want to make sure that the syntax will not be at an expense of ES5 users.

**Comment by robwormald** [comment]
aside (probably another issue needs to be created for this) but it would be suuuper cool if we could have an Observable<change> interface for this, something like
```
@ObserveChanges()
constructor(){
  this.onChanges.subscribe((changes) => {});
}
```

If the above was done with proper Observables (vs Subject/EventEmitter), it would be significantly lighter, as they would remain cold unless subscribed.

Just a thought.

**Comment by vicb** [comment]
+1.

There could be a migration path using name import `@Lifecycle.OnChange`:
- Create the namespace, promote its usage `@Lifecycle.OnChange` vs `OnChange` interface. Deprecate the former interfaces
- Remove the interfaces,
- Deprecate the `Lifecycle` namespace, make the decorators available without prefix,
- Remove the namespace.

**Comment by ericmartinezr** [comment]
+1 and feature request ?

My two cents. So far we put everything that we can run on component's initialization inside `onInit()`, so we end up with a huge method with a lot code. I see this proposal as 1:1 (1 decorator : 1 method), I think it would be nice to make it 1:N (1 decorator (per method) : N methods) and then every method would be concatenated.

```
directives : [Cmp1, Cmp2]

@OnInit()
DoSomethingWithCmp1() {}

@OnInit()
DoSomethingWithCmp2() {}
```

I don't know if it's a good idea or possible, but I had this idea and wanted to write it down :)

**Comment by vicb** [comment]
@ericmartinezr I don't think multiple `@OnInit` is a good idea:
- It would make the code less readable,
- It would make the code less predicate (what about the order in which methods are called),
- It would make it harder to test (you would have to update your tests when you add/remove `@OnInit`),
- It would probably make Angular (a tiny bit) less efficient.

On top of that it's easy to call as many methods as you want from any hook.

**Comment by btford** [comment]
@ericmartinezr I think it's better to use functional composition for running multiple methods on a single hook. Makes refactoring easier and avoids surprises WRT ordering.

The case where one method has multiple decorators should work fine.

**Comment by dsebastien** [comment]
+1. Java developers are gonna feel at home.. ;)

**Comment by alexpods** [comment]
+1 for this feature.

I also have a question. Why it was never considered to use Symbols (or some randomly generated tokens for browsers that doesn't support it). For example @IgorMinar's example would look like:

```js
import {Component, onInit} from 'angular2/angular2';

@Component({
  selector: 'test-foo'
})
class TestFoo {
  [onInit]() { ... }
}
```
Just because it's not so syntactically cool as decorators? or there are some limitations to do it?

**Comment by webteckie** [comment]
While we are in the subject was it considered at some point to do all the configuration in the `@Component` decorator and let the class be free of decorators?  E.g. WRT OnInit (and note that this may make it possible to do what @ericmartinezr was suggesting since the array can maintain order -- but this is not the point I'm making which is to keep the class free of decorators):
```
import {Component, onInit} from 'angular2/angular2';

@Component({
  selector: 'test-foo',
  OnInit: [init1, init2, ...]
})
class TestFoo {
  init1() { ... }
  init2() { ... }
}
```

**Comment by pkozlowski-opensource** [comment]
@webteckie what would be the advantage of "keeping the class free of decorators"? The downside of the   `OnInit: [init1, init2, ...]` approach is that you are duplicating method names (so if I want to rename `init1` I need to do it in 2 places now).

**Comment by webteckie** [comment]
@pkozlowski-opensource it may be a naive suggestion on my part but from where I stand it would be advantageous to me if the component config was in one place; just like a configuration of a system is usually in a single place.  You quickly know where to go to see what the configuration is, instead of browsing a potentially significant number of lines to figure things out.  That would also keep the class to standard ES5/ES6 even if programming in Typescript or whatever.  Anyways, it's probably too late to be making suggestions like this; just wanted to say it :-)

**Comment by mhevery** [comment]
The `@Component` is the canonical place. ie `@Component({inputs})` is same as `@Input`, so the `@OnInit` should be same.

**Comment by jeffbcross** [comment]
@mhevery so you are saying that we should add a property to the `Component` decorator for each lifecycle hook? And would its value be a string, the name of the method that should be called for that hook?

```typescript
@Component({
  onInit: 'onInit'
})
```

The Symbol-based approach proposed by @alexpods is a good thought, with some downsides:
 * It wouldn't necessarily make for a very good experience for dart users, and would add some complexity to the framework to support both targets.
 * It would be one more language feature that users would need to understand in order to use angular.
 * The ES5 experience would force the user to either statically-assign symbol names on the prototype with dot notation (`MyComponent.prototype.$$AngularInternalSymbolName`), or use square brackets to assign instance methods inside the component's constructor, or on the component's prototype (`MyComponent.prototype[ng.onInit] = function() {...}`).

I think we should move forward with the method decorator approach originally proposed by @IgorMinar, with the migration plan proposed by @vicb.

**Comment by IgorMinar** [comment]
@jeffbcross let's not dismiss symbols too quickly. We should aim for the best possible api for a particular language, if symbols are it for es6/ts then we should use them. Otherwise the api will be mediocre.

I thought about symbols in the past, but dismissed them just like you did. @alexpods example clearly shows how much more elegant the api could be. Let's spend a bit more time thinking about what would it take to make that a thing.

**Comment by alexpods** [comment]
@jeffbcross Thanks for clarification! I don't think the point about ES5 is a good argument (we don't have decorators in ES5/6, so some workaround is needed anyway). But the first two points are pretty solid. 

@IgorMinar Actually, after some thinking I also more incline to use of decorators. My points:

- `angular2` already uses decorators pretty actively. API will be more consistent if it continue to use them (and only them). 
-  you can add parameters to decorators, which can be very useful in the future. For example:
```js
@Component({ /* ... */ })
class TestFoo {
  @OnInit('someParam') someMethod() { ... }
}
```

What I don't like is that if we restrict use of decorators only for one method in a class then most likely we'll  repeat a name of a decorator in a method name:

```js
@Component({ /* ... */ })
class TestFoo {
  @OnInit() onInit() {
  }
  @OnActivate() onActivate() {
  }
  @OnSomething() onSomething() {
  }
```

Looks stupid for me. So, I think, @ericmartinezr's proposal about multiple decorators is pretty good. What about order, I think, `angular2` should not guarantee any order. If you use multiple decorators, you should use them for absolutely independent tasks, so order doesn't have any matter here (otherwise use functional composition).

**Comment by jeffbcross** [comment]
> I don't think the point about ES5 is a good argument (we don't have decorators in ES5/6, so some workaround is needed anyway).

@alexpods valid point, we provide some API sugar to make Component decoration easier for ES5 users, and we could do the same for this case. It's just not yet clear in my head what the right design would be.

I think @IgorMinar is more of the `Symbol` persuasion now. We're going to have a quick chat internally today to talk about how we could make this a good experience for Dart and TS/JS.

**Comment by mhevery** [comment]
> @Component({
>   onInit: 'onInit'
> })

Yes something like that.

**Comment by IgorMinar** [comment]
There is a difference between our current use of property decorators and proposed use for lifecycle hooks.

Lifecycle hooks have these properties:
- they have a predefined type signature (each hook is called with arguments of a certain type)
- there is only one property for each hook
- and lastly the name of the property is insignificant as long as the framework can identify it

None of these are true for our current use of property decorators.

Symbols have two big advantages over property decorators:
- concise - no need to duplicate the name two times or create a bogus property name
- type-safe - along with symbols we can also provide interfaces that will describe the hook signature

The biggest question mark is how to reconcile the implementation that uses symbols in TS/ES6 with other implementations used for ES5 and Dart.

**Comment by IgorMinar** [comment]
@vsavkin @yjbanov and me talked this over and came to the following conclusions:

- symbols provide the most ergonomic api (there is a slight concern about perceived API style inconsistency, but we believe that the benefits of symbols (uniqueness, type-safety, extensibility to 3rd party hooks) outweigh this concern)
- another concern with hooks expressed via symbols was that they couldn't be parameterized (unlike property decorators), however we see that this can be achieved either by providing several symbols or decorating properties defined via symbols.
- we considered how to implement this in Dart and came to conclusion that in Dart we could stick to using interfaces, which are retained at runtime (unlike in TS), so interfaces in dart give us type safety, name-collision prevention as well as api style conciseness. so the path forward would be to keep declaring lifecycle hooks via interfaces in dart, but use symbols in TS/JS

For these reasons, we'd like to proceed with using symbols in TS/JS and interfaces in Dart.

There are four remaining concerns we need to investigate before committing to symbols:
- TS support (type checking, tooling as well as transpilation)
- the current status of polyfilling symbols
- verify that there are no perf penalties when using symbols (e.g. `doCheck` hook is perf sensitive)
- the ES5 api

If none of these prove to be problematic then we should proceed with using symbols in TS/JS.

**Comment by yjbanov** [comment]
:+1: but I wouldn't mind the original decorator/annotation-based proposal due to the advantage that it would simplify documentation.

**Comment by jeffbcross** [comment]
:+1: I'll get to researching the blockers

**Comment by jeffbcross** [comment]
> TS support (type checking, tooling as well as transpilation)

Turns out we can't use interfaces in conjunction with symbols (except for built-in symbols) because it can't be statically verified. `var s = Symbol('OnInit')` is dynamic. We will move forward without good interface support for TypeScript.

**Comment by jeffbcross** [comment]
> verify that there are no perf penalties when using symbols (e.g. doCheck hook is perf sensitive)

I created a [jsperf benchmark](http://jsperf.com/computed-properties) to compare performance of computed method names against static method names. The computed properties just get transpiled to `SomeComponent.prototype[computedProperty]`. The results are pretty much identical. I have one benchmark which also tests against using a real `Symbol` for the property, which is about 4% slower, but still fast.

![screen shot 2015-11-04 at 2 57 58 pm](https://cloud.githubusercontent.com/assets/463703/10954795/782ff8b4-8304-11e5-9a64-9e575f2a4801.png)

**Comment by jeffbcross** [comment]
> the current status of polyfilling symbols

ES6-shim doesn't provide a polyfill for symbol, and we don't currently include a polyfill in our angular2.js bundle. ES6-shim will create a Symbol object, which it will use internally, but it is not a function. 

This project looks reasonable: https://github.com/medikoo/es6-symbol. It's 9k minified, about 2k gzipped, MIT licensed. If that's more bytes than we can bear, perhaps a good first step is to just use a string or object, which we could transparently change to a `Symbol` later.

**Comment by mhevery** [comment]
@IgorMinar Interfaces in Dart and Symbols in ES6 looks reasonable, But I don't see how this will work in ES5? Before we go down this path, can we also discuss/design an es5 solution, which is easy to use. (ie symbols are not easy to use in ES5)

**Comment by mhevery** [comment]
My other question is, if this is required for beta? We could always introduce symbol support later. Since we have to support interfaces in dart, I would vote to delaying this until after beta as I don't see any advantages in solving this before beta.

**Comment by jeffbcross** [comment]
> the ES5 api

Here is a comparison of the TypeScript/ES6 API vs ES5.

TS/ES6
```typescript
import {Component, onInit} from 'angular2/angular2';
@Component(...)
class MyComponent {
  [onInit] () {
  }
}
```

ES5
```javascript
function MyComponent () {}
MyComponent.prototype[ng.onInit] = function() {}
ng.Component(...).Class(MyComponent)
```

CC @mhevery, @IgorMinar

**Comment by jeffbcross** [comment]
@mhevery this will be a breaking change, and impacts docs and usage, so should be done before beta or NEVER.

**Comment by mhevery** [comment]
@jeffbcross Actually that is not how they will declare them in ES5.

```
var MyComponent = Component({...}).Class({
  constructor: function() {...},
  ng.OnInit: function() {}. /// THIS DOES NOT WORK
})
```

Since we have to support interfaces, we could just stick with the current model. And then later introduces alternative way. The interfaces can not go away because of Dart. So it is not a breaking change, but rather additional way to do this.

**Comment by jeffbcross** [comment]
@mhevery I will let you and @IgorMinar work it out. I don't mind having less work to do :).

Users could _technically_ use `Object.defineProperty` with the current ES5 API, but that's not very nice. 

As a separate discussion, don't you think we should still support declaring prototype out-of-band? I.e. `.Class()` could take an object or a function?

**Comment by jeffbcross** [comment]
@mhevery we could also re-introduce the `lifecycle` property in the `Component` decorator.

**Comment by mhevery** [comment]
> As a separate discussion, don't you think we should still support declaring prototype out-of-band? I.e. .Class() could take an object or a function?

Can you provide an example?

**Comment by jeffbcross** [comment]
@mhevery 

```javascript
function MyComponent () {}
MyComponent.prototype[ng.onInit] = function() {}
ng.Component(...).Class(MyComponent)
```

**Comment by jeffbcross** [comment]
> Since we have to support interfaces, we could just stick with the current model. 

But we don't solve the problem with the current model, which is that a user could accidentally name a method in a way that collides with lifecycle hooks, which would then get called by angular unexpectedly.

**Comment by yjbanov** [comment]
Can we offer a React-style DSL for declaring components with lighter syntax?

**Comment by IgorMinar** [comment]
@mhevery interfaces don't work for JS/TS because they are not retained at runtime.

We currently just check if a well now method name is present on a class prototype and we call it, but that's error prone and can result in major headaches when we introduce more lifecycle hooks or when ionic and similar solutions on top of Angular introduce custom hooks (we already need custom hooks for the router and I expect that there will be more).

we control the ES5 DSL so we can say that  some properties are special (we already do this for `constructor` property). So for example we can say that `Class` dsl properties starting with `$` will be turned into lifecycle hooks:

```
var MyComponent = Component({...}).Class({
  constructor: function() {...},
  $onInit: function() {}. /// everything prefix with $ will be turned into `[ng.onInit]` under the hood.
})
```

**Comment by alexpods** [comment]
How about:
```js
var SomeComponent = ng
    .Component({
        selector: 'some-compnent',
        lifecycle: { method: ng.OnInit, anotherMethod: router.OnActivate }
    })
    .Class({
        method: function() {
            // code for  ng.OnInit lifecycle hook
        },
        anotherMethod: function() {
            // code for router.OnActivate lifecycle hook 
        }
    })
```
This can work both for symbols and decorators.
Cons:
  - a little bit verbose
  - most likely names of lifecycle hooks will be repeated: `lifecycle: { onInit: ng.OnInit, onActivate: ng.onActivate }`

@IgorMinar 
>  $onInit: function() {}. /// will be turned into `[ng.onInit]` under the hood

But you can't turn it into `[router.onInit]` or `[animation.onStart]` or `[ionic.somHook]`. So it doesn't work for third-party library.

**Comment by mhevery** [comment]
@jeffbcross https://github.com/angular/angular/issues/5036#issuecomment-153910297

```
function MyComponent () {}
MyComponent.prototype[ng.onInit] = function() {}
ng.Component(...).Class(MyComponent)
```

Yes we could support the above, but it would be very awkward. The whole point of `Class` is to make it streamline, and this breaks it up.

**Comment by pkozlowski-opensource** [comment]
> the current status of polyfilling symbols

While researching potential polyfills could we test them in all the supported browsers? 

> My other question is, if this is required for beta? 

I hate to leave comments / feedback like this, but personally I would love that we address more pressing pbs for beta (my personal opinion).

**Comment by robwormald** [comment]
Note we are running into some issues with Symbols coming from RxJS https://github.com/angular/angular/issues/5072 might not be a blocker, but be aware. @mlaval might have more input.

**Comment by mhevery** [comment]
> @mhevery interfaces don't work for JS/TS because they are not retained at runtime.

It depends what you mean by retained. In Dart interfaces are retained nominally, but in JS they are structural. So it does work as intended. It is very common in JS to say something like this:
```
if (obj.foo) obj.foo()
```

Which is what structural typing is all about, so it does work as intended. BWT the only difference between what we have now, and symbols is that symbols can not be guessed hence prevent collision. But we could achieve the same by prefixing it with `ng` or `$` or some other unique prefix.

My concern is that we now have three ways of declaring a call back. In Dart you use interfaces, in TS/ES6 you would use symbols and in ES5 you would use `$` prefix. This is not a good situation to be in. On top of that we pay 9kb (2kb gzip) on already large codebase. This creates a lot more problems then it solves. (Also now you have to explain symbols to ES5 developers, to further confuse the situation)

On top of that the proposal breaks type safety (and tooling). For example let's say you would like to compile this with JSCompiler,  there will be no way to express these types in a way which JS Compiler understands.

My current stand is that we should create consistent experience across all languages, not fragment it. This leaves us with these options.

1) Use interfaces as now. Downside is that there could be name collision, we could solve this with prefixing the methods. Also extensibility is not clear. 
2) Use annotation/decorators. Downside is that it is a bit more verbose but it is consistent across all languages. (Mostly, in Dart and TypeScript it would be strongly advised to have `extends` clause to have proper type safety.)

Dart / TypeScript
```
@Component()
class MyClass extends OnInit {
  @OnInit onInit() {...}
}
```

ES5
```
var MyClass = Component().Class({
  onInit: [OnInit(), function() {...}]
})
```

Given that `extends` clause is a good idea from the type safety perspective, I think we should just stick with it. Yes, there are problems with collision, but those are mostly theoretical and will easily be detected with tolls, since the method signatures will most likely not match.

**Comment by pkozlowski-opensource** [comment]
For what it is worth, I'm with @mhevery here - we've got already nice amount of new concepts / things that contribute to the size of the code base. IMO symbol-based API doesn't pays for its price.

**Comment by mhevery** [comment]
# COUNTER PROPOSAL

I have done some more thinking and I think I can provide a counter proposal which satisfy all of the constraints.

First let's agree that Dart does not have a problem. That is:
1) Interfaces provide right level of type safety.
2) Nominal `instanceof` does prevent name collision form accidentally implementing a interface.
3) It is easy for third parties such as router to have their own interfaces.

The solution is to just fix JS types from structural to nominal. Here is how it would work.

```
@Component({
  implements: [OnInit] // << Make nominal types explicit
})
class MyComponent implements OnInit {
  onInit() { /* works as intended */ },
  onDestroy() { /* does not get confused because it is not listed in the decoration. */  }
}
```

Third parties:

```
// Third Party can declare an interface
class OnActivete { // must be class so that it has runtime representation
  onActivate() {}
}

@Component({
  implements: [OnActivate]
})
class MyComponent implements OnActivate {
  onActivate() {  /* Works as expected */}
}
```

we then provide a method such as 

```
isNominallyInstanceOf(new MyComponent(), OnActivate) => for instance checks 
isNominallyImplements(MyComponent, OnActivate) => for type checks
```

Notes:
- `implements` on `Component` would be JS only. IE, in Dart we would no need it.

**Comment by e-oz** [comment]
I found this issue after reading first time about onInit and siblings (I like this new thing!) and tried to submit issue with idea how to simplify their declaration (because of collisions, mentioned by Igor).

My idea is use Dependency Injection:

```
class SomeComponent(){
  constructor(SomeDependency, LifecycleHooks){
    LifecycleHooks.onInit(() => { /* handler code here */});
  }
}
```

What do you think?

**Comment by alexpods** [comment]
@e-oz It doesn't work. You must have an ability to check existence of lifecycle hooks in a Type, not only in its instance.

**Comment by pkozlowski-opensource** [comment]
@mhevery while your latest proposal works it kind of duplicates the info and IMO it will be constant source of confusion (hay, I've implemented and interface but my lifecycle hook still doesn't work...).

Do we really have such huge problems (I mean, in practice) with decorators?

**Comment by e-oz** [comment]
@mhevery 
> 2) Nominal instanceof does prevent name collision form accidentally implementing a interface.

It doesn't prevent users to declare some methods, which will be reserved by lifecycle hooks in future versions of Angular.

@alexpods thanks, didn't know that. Well, then I have no better ideas :)

**Comment by IgorMinar** [comment]
@mhevery as @pkozlowski-opensource your conter-proposal is even more verbose than using just property decorators, adds more confusion and doesn't have any new benefits. Am I missing something?

**Comment by robwormald** [comment]
All of these seems like a whole lot of work and complexity to solve a problem which is far easier solved with documentation, IMO. React does what we are doing currently, and as far as I'm aware the vast majority of their community does just fine with it. Add in the benefits of `implements` and I just don't see the value. 

https://facebook.github.io/react/docs/component-specs.html

**Comment by robwormald** [comment]
Looking at those react docs, the longer `componentOnWhatever` method names seem like they'd prevent most collisions.
`componentOnInit`
`componentOnDestroy`
etc

**Comment by IgorMinar** [comment]
@robwormald convention and docs are not sufficient. we need an extensible mechanism that can be used by 3rd parties. keep in mind that some apps we are building will have millions of lines of code. we can't rely on documentation to prevent name collisions.

**Comment by mhevery** [comment]
@e-oz
> 2) Nominal instanceof does prevent name collision form accidentally implementing a interface.
> It doesn't prevent users to declare some methods, which will be reserved by lifecycle hooks in future versions of Angular.

It does, since I can declare a same method and it will be mine. When I go and implement the interface I will be forced to rename my method. The point is that I can't accidentally declare a method and then framework would call it.

**Comment by mhevery** [comment]
> @mhevery as @pkozlowski-opensource your conter-proposal is even more verbose than using just property decorators, adds more confusion and doesn't have any new benefits. Am I missing something?

The new benefit is that it does not have the problem of accidental name collision. If user declares a new method `onTick` and then framework choses to take over `onTick` in future there is no collision since you don't have `OnTick` in the `implements` decoration. So `onTick` is user method not framework. If user the choses to implement `OnTick` by adding it to `implements` only then does she have to resolve the name collision. It is not accidental, it is explicit. 

1) You have to declare `class MyClass implements OnInit` or you don't have type information. This part is not option, hence required.
2) (I personally think this point is an overkill, but) You have to somehow explicitly disambiguate `onInit` from user and framework namespace. There are several ways to do it.
  - a) As decorator `@OnInit onIni() {}`
  - b) As annotation in `@Component` such as `implements: [OnInit]` 
  - c) As a prefix `ngOnInit() {}` or `$onInit() {}`
  - d) Using symbol as `[OnInint.method]() {}`

I have already mentioned why Symbols are not a good idea here: https://github.com/angular/angular/issues/5036#issuecomment-154118691

**Comment by IgorMinar** [comment]
`@OnInit someName() {}` is sufficient to prevent name collisions. no need for pseudo-interface.

**Comment by e-oz** [comment]
@mhevery in other words, it prevents usage of interface when method with such name is already exist. So we can leave it as is (although I'm not sure how it will work in ES5/6 without interfaces), but user will have to do deep refactoring if case of such collision. Can't call it elegant solution, especially for libraries, when changes in API can be costly.

**Comment by e-oz** [comment]
(removed comment)
sorry, I forgot all APIs of components are encapsulated by inputs/outputs lists.

**Comment by bfanger** [comment]
Adding the `@Component` decorator to a class is an explicit developer action, which in my opinion is an opt-in to the lifecycle methods.

It also means that if the existing code has a namespace collision, it can be changed/fixed by writing either a wrapper class or changing the the original code.

Adding additional boilerplate (annotations) for the common scenario should be prevented if possible.

**Comment by timkindberg** [comment]
My 2 cents: Sounds like `@OnInit init() {}` is the best compromise. Not overly verbose, explicit, consistent dev experience, avoids lame strings in @Component property (though probably you should put the property in there as well).

**Comment by robwormald** [comment]
Was building a few custom components this evening, and (possibly because I'm doing Rx all the time) it feels like there's a mismatch:

We use Observable/EventEmitter to notify outside the component that something has changed, but the inverse (this discussion) we use regular callback methods. 

To me, something like:
```typescript

class MyComponent {
  @OnInit init : Observable<any>
  @OnChange didChange : Observable<ChangeLog>
  constructor(){
    this.init.subscribe(_ => console.log('intialized!'));
    this.didChange.subscribe(changes => {
        //do stuff with changes
     });
    //enables more stuff like
    this.didChange.filter(change => change.field === 'foo').subscribe(...);
    this.didChange.debounceTime(200).subscribe(...);
  }
}

```
makes more sense. No?

**Comment by wesleycho** [comment]
@robwormald pointed me to this thread since I had a question about `onChange` currently catching all updates to input properties when working on ng-bootstrap - currently, the only way I know of to watch any input change is via the `onChange` lifecycle hook. However, this becomes problematic as soon as one has to watch for updates to more than one attribute.

I would like a syntax that supports watching to an attribute, such as
```js
@Component({
  inputs: ['attribute1', 'attribute2']
})
class MyComponent {
  @OnChange('attribute1') onAttr1Change: Observable<Attr1Type>;
  @OnChange('attribute2') onAttr2Change: Observable<Attr2Type>;
}
```
This would give the consumer the ability to observe multiple attributes

**Comment by timkindberg** [comment]
@wesleycho it's funny you bring that idea up. I was initially thinking (and even at one point bugging @robwormald with questions about) how I thought inputs would be observables by default. It kind of makes sense since they are values that change. Our outputs are Observables... so why not inputs. When I heard Angular 2 was going to bake Rx in, I really thought it meant *really bake it in*. But it seems to sort of live just here or there in the api.

I like @wesleycho's suggestion, however I could see someone wanting an even shorter-hand way of declaring an observable input. Maybe (ignore name):

```ts
@Component({})
class MyComponent {
  @rxInput() attr1Changes: Observable<Attr1Type>;
}
```

However, this isn't that great, because the whole point of making it an observable is so you can perform methods on it (e.g. map, reduce, etc). So we'd still have to do that in a class method somewhere. You could do it in the constructor, but now we introduce a new variable to store the results of the subscription's next values.

```ts
@Component({})
class MyComponent {
  @rxInput() attr1Changes: Observable<Attr1Type>;
  constructor() {
    // We either have to set it up and subscribe, creating a new annoying val to store the results
    this.attr1Changes.map().reduce().subscribe((val) -> this.attr1Val = val);

    // or we have to store the new transformed observable into a new val anyway, to be used with
    // the async filter
    this.attr1ChangesTransformed = this.attr1Changes.map().reduce();
  }
}
```

To sum up:

**The Problem:**
I want to declare an input as an observable of changes, and put those changes on my view with as little work as possible. Any more ideas? 

@wesleycho, could you show what steps you'd take to get your onAttr1Change next values onto the view? Assume you still need to run that observable through some more rx operators.

**Comment by wesleycho** [comment]
My idea was originally to do
```js
class MyComponent {
  constructor() {
    this.onAttr1Change.subscribe(v => this.attr1 = v);
  }
}
```
Or something like that - in the use case I was thinking, there actually ended up being a much better way to design my component using QueryList.

**Comment by e-oz** [comment]
Idea about Observables for lifecycle hooks looks very attractive.

**Comment by jeffbcross** [comment]
New API was decided after much debate between @mhevery and @IgorMinar: Just prefix method names with `ng`. I.e. `ngAfterInit`. This is contrary to my proposal of using shorter, more unique names for lifecycle methods* like `login`, `logout`, `create`, `remove`, `update`, `get`, `set`, but you gotta pick your battles :/.

I'm implementing the prefix fix now.

*this is a joke, but the prefix decision is real.

**Comment by jeffbcross** [comment]
@vsavkin & @IgorMinar I think I should prefix pipe's `onDestroy` for consistency: `pipeOnDestroy`. Agree/disagree?

**Comment by e-oz** [comment]
@jeffbcross prefix is `ng` or `pipe`?

**Comment by jeffbcross** [comment]
@e-oz `pipe`, consistent with `router` lifecycle hooks like `routerCanActivate`

**Comment by e-oz** [comment]
thanks

**Comment by jeffbcross** [comment]
Talked with @vsavkin and he made the point that directives implement `OnDestroy`/`ngOnDestroy`, so it would be most consistent for pipes to just implement the angular lifecycle interface. So I'll change pipe instances to implement `OnDestroy` and will rename method to `ngOnDestroy`.

**Comment by jeffbcross** [comment]
Update: we decided to wait a little bit to land this change since it's be a lot of work to merge internally.

**Comment by mhevery** [comment]
It should be `ngOnDestroy` not `pipeOnDestroy` because it is coming from the `ng` library. We don`t have `componenOnDestroy` and `pipeOnDestroy` its just `ngOnDestroy`. Also we can just have a single interface as already pointed out.

**Comment by jeffbcross** [comment]
@mhevery I think what you're saying is consistent with my last comment on the matter, right?
